vowpalwabbit/csoaa.cc

```diff
@@ -904,18 +833,61 @@ base_learner* csldf_setup(arguments& arg)
     pred_type = prediction_type::multiclass;
 
   ld->read_example_this_loop = 0;
-  ld->need_to_clear = false;
-  ldf* bare = ld.get();
-  learner<ldf>& l = init_learner(ld, setup_base(arg), predict_or_learn<true>, predict_or_learn<false>, 1, pred_type);
-  if (bare->is_singleline)
-    l.set_finish_example(finish_singleline_example);
-  else
-    l.set_finish_example(finish_multiline_example);
+  learner<ldf,multi_ex>& l = init_learner(ld, setup_base(arg), do_actual_learning<true>, do_actual_learning<false>, 1, pred_type);
+  l.set_finish_example(finish_multiline_example);
   l.set_finish(finish);
-  l.set_end_examples(end_examples);
   l.set_end_pass(end_pass);
+  l.set_test_example(COST_SENSITIVE::example_is_test);
   arg.all->cost_sensitive = make_base(l);
   return arg.all->cost_sensitive;
 }
 
+/*
```

Comment:
Shift up so setup_example() is at bottom of file (in canonical position).

Reply:
Todo: minor

Reply:
Not sure what this means.  I don't see a setup_example() in the code.

Reply:
Done.